spec/formatting.md

```diff
@@ -18,3 +18,192 @@ the local variable takes precedence.
+    does not equal the number of Selectors.
+
+    Example invalid messages resulting in a Variant Key Mismatch error:
+
+    ```
+    match {$one}
+    when 1 2 {Too many}
+    when * {Otherwise}
+    ```
+
+    ```
+    match {$one} {$two}
+    when 1 2 {Two keys}
+    when * {Too few}
+    when * * {Otherwise}
+    ```
+
+  - **Missing Fallback Variant errors** occur when the message
+    does not include a Variant with only catch-all keys.
+
+    Example invalid messages resulting in a Missing Fallback Variant error:
+
+    ```
+    match {$one}
+    when 1 {Value is one}
+    when 2 {Value is two}
+    ```
+
+    ```
+    match {$one} {$two}
+    when 1 * {First is one}
+    when * 1 {Second is one}
+    ```
+
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable errors** occur when a variable reference cannot be resolved.
+
+    For example, attempting to format either of the following messages
+    must result in an Unresolved Variable error if done within a context that
+    does not provide for the variable reference `$var` to be successfully resolved:
+
+    ```
+    {The value is {$var}.}
+    ```
+
+    ```
+    match {$var}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+- **Selection errors** occur when message selection fails.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
+
+    For example, attempting to format either of the following messages
+    may result in a Selector error if done within a context that
+    uses a `:plural` selector function which requires its input to be numeric:
+
+    ```
+    match {(horse) :plural}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+    ```
+    let $sel = {(horse) :plural}
+    match {$sel}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+- **Formatting errors** occur during the formatting of a resolved value,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+  For example, attempting to format any of the following messages
+  may result in a Formatting error if done within a context that
+
+  1. provides for the variable reference `$user` to resolve to
+     an object `{ name: 'Kat', id: 1234 }`, and
+  2. uses a `:get` formatting function which requires its argument to be an object and
+     an option `field` to be provided with a string value,
+
+  ```
+  {Hello, {(horse) :get field=name}!}
+  ```
+
+  ```
+  {Hello, {$user :get}!}
+  ```
+
+  ```
+  let $id = {$user :get field=id}
+  {Hello, {$id :get field=name}!}
+  ```
+
+Syntax and Data Model errors must be emitted as soon as possible.
+
+During selection, an Expression handler must only emit Resolution and Selection errors.
+During formatting, an Expression handler must only emit Resolution and Formatting errors.
```

Comment:
Same question as above: This phrasing doesn't explicitly mandate that errors should chain (or nest, depending how you look at it). I.e a resolution error can lead to a formatting error. Implementations should emit both to enable higher-level abstractions to react differently to these different failures.